In [3]:
import pandas as pd

# Load the dataset
df = pd.read_csv("Dataset 2.csv")

# Q1: Display the first five records
print("First Five Records:")
print(df.head())

# Q2: Determine the number of rows and columns
rows, columns = df.shape
print("\nNumber of Rows:", rows)
print("Number of Columns:", columns)

# Q3: Display all column names
print("\nColumn Names:")
print(df.columns.tolist())

# Q4: Identify numerical and categorical features
numerical_features = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_features = df.select_dtypes(include=['object']).columns.tolist()

print("\nNumerical Features:")
print(numerical_features)

print("\nCategorical Features:")
print(categorical_features)

# Q5: Check for missing values
print("\nMissing Values in Each Column:")
print(df.isnull().sum())

# Check if any missing values exist
if df.isnull().values.any():
    print("\nThe dataset contains missing values.")
else:
    print("\nThe dataset does not contain missing values.")

First Five Records:
   UserID  Age  Gender SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22  Female            Basic                 23            5   
1    1002   55    Male            Basic                  9            4   
2    1003   49    Male            Basic                  8            3   
3    1004   39  Female          Premium                 19            5   
4    1005   38  Female          Premium                 23            5   

  FavoriteGenre  AdClicks  MonthlySpend SubscriptionRenewed  
0        Comedy        13           353                  No  
1         Drama        14           317                 Yes  
2        Comedy        16           309                  No  
3         Drama        45           833                 Yes  
4        Sci-Fi        24           804                 Yes  

Number of Rows: 750
Number of Columns: 10

Column Names:
['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre'

In [4]:
#Q6: average age of users
average_age = df['Age'].mean()
print("\nAverage Age of Users:", average_age)


Average Age of Users: 41.824


In [6]:
#Q7: average watch hours per week
average_watch_hours = df['WatchHoursPerWeek'].mean()
print("\nAverage Watch Hours per Week:", average_watch_hours)
#


Average Watch Hours per Week: 14.236


In [7]:
#Q8: average monthly spending of users
average_monthly_spending = df['MonthlySpend'].mean()
print("\nAverage Monthly Spending:", average_monthly_spending)



Average Monthly Spending: 689.9053333333334


In [8]:
#Q9: the number of users in each subscription category
subscription_counts = df['SubscriptionType'].value_counts()
print("\nNumber of Users in Each Subscription Category:")
print(subscription_counts)


Number of Users in Each Subscription Category:
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [10]:
#Q10: the percentage of users who renewed their subscriptions
#percentage of yes in SubscriptionRenewed
yes_count = df['SubscriptionRenewed'].value_counts().get('Yes', 0)
total_count = len(df)
yes_percentage = (yes_count / total_count) * 100
print(yes_percentage)


46.266666666666666


Data Preparation

In [11]:
#Q11: Convert categorical features into numerical form


# Converts text columns into numerical values
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['SubscriptionType'] = le.fit_transform(df['SubscriptionType'])
df['SubscriptionRenewed'] = le.fit_transform(df['SubscriptionRenewed'])
df['Gender'] = le.fit_transform(df['Gender'])
df['FavoriteGenre']= le.fit_transform(df['FavoriteGenre'])


In [12]:
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,0,0,23,5,1,13,353,0
1,1002,55,1,0,9,4,2,14,317,1
2,1003,49,1,0,8,3,1,16,309,0
3,1004,39,0,1,19,5,2,45,833,1
4,1005,38,0,1,23,5,5,24,804,1


In [13]:
#Q12: Define the feature set (X) and target variable (y) for subscription renewal prediction

X = df.drop(['UserID','SubscriptionRenewed'], axis=1)

y = df['SubscriptionRenewed']

In [15]:
#Q13: Split the dataset into training and testing sets.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

Decision Tree Classification

In [ ]:
#Q14: Train a Decision Tree model to predict whether a user will renew their subscription.
#Q15: Evaluate the model accuracy.
#Q16: Generate and interpret the confusion matrix.

In [17]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

pred = dt.predict(X_test)

from sklearn.metrics import accuracy_score, confusion_matrix

print("Decision Tree Accuracy:", accuracy_score(y_test,pred))

print('\nConfusion Matrix')
print(confusion_matrix(y_test,pred))

Decision Tree Accuracy: 0.56

Confusion Matrix
[[48 34]
 [32 36]]


K-Nearest Neighbors (KNN)

In [18]:
#Q17: Train a KNN classifier with K=5.

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)

knn.fit(X_train,y_train)

knn_pred = knn.predict(X_test)

print("KNN Accuracy:",
      accuracy_score(y_test,knn_pred))

KNN Accuracy: 0.6333333333333333


In [19]:
#Q18: Compare the accuracy of KNN with the Decision Tree model.

dt_acc = accuracy_score(y_test,pred)

knn_acc = accuracy_score(y_test,knn_pred)

print("Decision Tree:",dt_acc)
print("KNN:",knn_acc)

if dt_acc > knn_acc:
    print("Decision Tree performed better")
else:
    print("KNN performed better")

Decision Tree: 0.56
KNN: 0.6333333333333333
KNN performed better


Linear Regression

In [20]:
#Q19: Train a Linear Regression model to predict monthly spending.
from sklearn.linear_model import LinearRegression

X_reg = df.drop(['UserID', 'MonthlySpend'], axis=1)

y_reg = df['MonthlySpend']

X_train_reg,X_test_reg,y_train_reg,y_test_reg = train_test_split(
    X_reg,y_reg,test_size=0.2,random_state=42
)

lr = LinearRegression()

lr.fit(X_train_reg,y_train_reg)

LinearRegression()

In [27]:
#Q20: Predict the monthly spending for a new user and interpret the result.

new_user= [[21,0,1,10,3,1,11,0]]

predicted_spending = lr.predict(new_user)

print("Predicted Monthly Spending:",predicted_spending[0])




Predicted Monthly Spending: 809.526205678354


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Business Reflection Questions

1.Which factors appear to influence subscription renewal the most?

Ans. Likely influential factors include:
* WatchHoursPerWeek
* MonthlySpend
* SubscriptionType
* DevicesUsed
* AdClicks

Users who watch more content and spend more money are generally more likely to renew.

2.Why is subscription renewal a classification problem?

Ans. Because the target variable has two categories:
* Yes
* No

The model predicts a class label rather than a numeric value.

3.Why is monthly spending a regression problem?

Ans. Monthly spending is a continuous numerical value (₹), so we predict a number rather than a category.

4.Which algorithm performed better?

Ans. KNN (63.33%) performed better than Decision Tree (56%).

5.How can Netflix use these predictions?

Ans.

* Identify users likely to cancel.

* Offer personalized recommendations.

* Provide discounts or loyalty rewards.

* Improve customer retention campaigns.

* Increase revenue through targeted marketing.